core/cog-gamepad-manette.c

```diff
@@ -0,0 +1,325 @@
+{
+    uint16_t button;
+    if (!manette_event_get_button(event, &button))
+        return;
+
+    wpe_gamepad_dispatch_button_changed(gamepad->server, toStandardGamepadButton(button), false);
+}
+
+static void
+device_disconnect_cb(ManetteDevice *device, Gamepad *gamepad)
+{
+    if (!gamepad)
+        return;
+
+    if (gamepad->button_pressed > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->button_pressed);
+        gamepad->button_pressed = 0;
+    }
+
+    if (gamepad->button_released > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->button_released);
+        gamepad->button_released = 0;
+    }
+
+    if (gamepad->axis_event > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->axis_event);
+        gamepad->axis_event = 0;
+    }
+
+    if (gamepad->disconnected > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->disconnected);
+        gamepad->disconnected = 0;
+    }
+}
+
+static gpointer
+gamepad_create(struct wpe_gamepad *server, unsigned gamepad_id)
+{
+    Gamepad       *gamepad;
+    ManetteDevice *device;
+
+    device = GUINT_TO_POINTER(gamepad_id);
+    if (!MANETTE_IS_DEVICE(device))
+        return NULL;
+
+    gamepad = g_new0(Gamepad, 1);
+
+    gamepad->server = server;
+    gamepad->device = g_object_ref(device);
+
+    gamepad->button_pressed = g_signal_connect(device, "button-press-event", G_CALLBACK(on_button_press), gamepad);
+    gamepad->button_released = g_signal_connect(device, "button-release-event", G_CALLBACK(on_button_release), gamepad);
+    gamepad->axis_event = g_signal_connect(device, "absolute-axis-event", G_CALLBACK(on_absolute_axis), gamepad);
+    gamepad->disconnected = g_signal_connect(device, "disconnected", G_CALLBACK(device_disconnect_cb), gamepad);
+
+    g_message("gamepad: %s - created", manette_device_get_name(device));
+
+    return gamepad;
+}
+
+static void
+gamepad_destroy(gpointer data)
+{
+    Gamepad *gamepad = data;
+
+    if (!(gamepad && gamepad->device))
+        return;
+
+    g_clear_object(&gamepad->device);
+    g_free(gamepad);
+}
+
+static const char *
+gamepad_get_id(gpointer data)
+{
+    Gamepad *gamepad = data;
+
+    if (!(gamepad && gamepad->device))
+        return "Unknown";
+
+    return manette_device_get_name(gamepad->device);
+}
+
+static struct wpe_gamepad_provider_interface provider_interface = {
+    .create = provider_create,
+    .destroy = provider_destroy,
+    .start = provider_start,
+    .stop = provider_stop,
+};
+
+static struct wpe_gamepad_interface device_interface = {
+    .create = gamepad_create,
+    .destroy = gamepad_destroy,
+    .get_id = gamepad_get_id,
+};
+
+void
+cog_register_gamepad_backend(GamepadProviderGetViewBackend *provider_get_view_for_gamepad)
+{
+    provider_interface.get_view_backend = provider_get_view_for_gamepad;
+    wpe_gamepad_set_handler(&provider_interface, &device_interface);
```

Comment:
I'm not a fan of this, because it means that there can only ever be one gamepad backend, and it has to be the Manette one. What if we wanted to build more than one and allow choosing at runtime? I think we can leave it as-is for now, but knowing that this may need changes later on.

Reply:
It's a nice feature to have, but perhaps, as you mention, not a priority.